File: src/lib/services/file.ts
// File operation service - communicates with Rust backend
import { invoke } from '@tauri-apps/api/core';

/**
 * Open file using file picker dialog
 * @returns Tuple of [filePath, content] or null if cancelled
 */
export async function openFileDialog(): Promise<[string, string] | null> {
  const result = await invoke<[string, string] | null>('open_file_dialog');
  return result;
}

/**
 * Save content to existing file path
 */
export async function saveFile(path: string, content: string): Promise<void> {
  await invoke('save_file', { path, content });
}

/**
 * Save content to new file using save dialog
 * @returns File path or null if cancelled
 */
export async function saveFileDialog(content: string, defaultFileName?: string): Promise<string | null> {
  // Tauri command args are deserialized from camelCase keys
  const payload: { content: string; defaultFileName?: string } = { content };
  if (defaultFileName) payload.defaultFileName = defaultFileName;
  const result = await invoke<string | null>('save_file_dialog', payload);
  return result;
}

/**
 * Read file content by path (for drag & drop)
 */
export async function readFile(path: string): Promise<string> {
  return await invoke<string>('read_file', { path });
}

/**
 * Check if file path is valid JSON file
 */
export async function isJsonFile(path: string): Promise<boolean> {
  return await invoke<boolean>('is_json_file', { path });
}

/**
 * Get file name from path
 */
export async function getFileName(path: string): Promise<string | null> {
  return await invoke<string | null>('get_file_name', { path });
}

File: src-tauri/src/commands/file.rs
// File operation commands
use std::path::PathBuf;
use tauri::AppHandle;
use tauri_plugin_dialog::DialogExt;

fn sanitize_file_name(input: &str) -> String {
    let trimmed = input.trim();
    let mut out = String::with_capacity(trimmed.len());

    for ch in trimmed.chars() {
        // Drop ASCII control chars
        if ch.is_control() {
            continue;
        }

        // Replace common illegal filename characters (esp. Windows) + path separators
        match ch {
            '<' | '>' | ':' | '"' | '/' | '\\' | '|' | '?' | '*' => out.push('_'),
            _ => out.push(ch),
        }
    }

    // Windows compatibility: filenames cannot end with space or dot
    let out = out.trim_end_matches([' ', '.']);

    // Avoid extremely long names
    out.chars().take(200).collect::<String>()
}

fn ensure_json_extension(name: &str) -> String {
    if name.to_lowercase().ends_with(".json") {
        name.to_string()
    } else {
        format!("{}.json", name)
    }
}

/// Open a JSON file using file picker dialog
#[tauri::command]
pub async fn open_file_dialog(app: AppHandle) -> Result<Option<(String, String)>, String> {
    let file_path = app
        .dialog()
        .file()
        .add_filter("JSON Files", &["json"])
        .add_filter("All Files", &["*"])
        .blocking_pick_file();

    match file_path {
        Some(path) => {
            let path_str = path.to_string();
            let path_buf = PathBuf::from(&path_str);
            match tokio::fs::read_to_string(&path_buf).await {
                Ok(content) => Ok(Some((path_str, content))),
                Err(e) => Err(format!("Failed to read file: {}", e)),
            }
        }
        None => Ok(None), // User cancelled
    }
}

/// Save content to a file (existing file path)
#[tauri::command]
pub async fn save_file(path: String, content: String) -> Result<(), String> {
    tokio::fs::write(&path, content)
        .await
        .map_err(|e| format!("Failed to save file: {}", e))
}

/// Save content to a new file using save dialog
#[tauri::command]
pub async fn save_file_dialog(
    app: AppHandle,
    content: String,
    default_file_name: Option<String>,
) -> Result<Option<String>, String> {
    let suggested = default_file_name.unwrap_or_else(|| "Untitled.json".to_string());
    let sanitized = sanitize_file_name(&suggested);
    let base = if sanitized.is_empty() {
        "Untitled".to_string()
    } else {
        sanitized
    };
    let final_name = ensure_json_extension(&base);

    let file_path = app
        .dialog()
        .file()
        .add_filter("JSON Files", &["json"])
        .add_filter("All Files", &["*"])
        .set_file_name(&final_name)
        .blocking_save_file();

    match file_path {
        Some(path) => {
            let path_str = path.to_string();
            let path_buf = PathBuf::from(&path_str);
            tokio::fs::write(&path_buf, content)
                .await
                .map_err(|e| format!("Failed to save file: {}", e))?;
            Ok(Some(path_str))
        }
        None => Ok(None), // User cancelled
    }
}

/// Read file content by path (for drag & drop)
#[tauri::command]
pub async fn read_file(path: String) -> Result<String, String> {
    tokio::fs::read_to_string(&path)
        .await
        .map_err(|e| format!("Failed to read file: {}", e))
}

/// Check if file path is valid JSON file
#[tauri::command]
pub fn is_json_file(path: String) -> bool {
    let path = PathBuf::from(path);
    path.extension()
        .and_then(|ext| ext.to_str())
        .map(|ext| ext.eq_ignore_ascii_case("json"))
        .unwrap_or(false)
}

/// Get file name from path
#[tauri::command]
pub fn get_file_name(path: String) -> Option<String> {
    PathBuf::from(path)
        .file_name()
        .and_then(|name| name.to_str())
        .map(|s| s.to_string())
}
